<a href="https://colab.research.google.com/github/yerkogallardo/perros_gatos_CNN/blob/main/Taller_Red_neuronal_simple.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Taller 1 de Redes Neuronales** 🧠💪🏼⚡

### **Parte práctica**

Esto es un Notebook de desarrollo, es una de las herramientas que los científicos de datos usamos de forma cotidiana para trabajar un modelo de Machine Learning o una Red Neuronal.  En este caso, es un notebook en la plataforma Google Colab, que nos permite sin necesidad de instalar nada en nuestro computador trabajar con Python y las librerias con las que vamos a realizar este taller.

### **Instrucciones:**

En este notebook encontrarás todo el código para montar una red neuronal capaz de encontrar el valor del siguiente número Fibonacci

En cada celda de código tendrás una parte para que puedas rellenar código y algunas donde podrás cambiar a tu gusto.

In [1]:
## Importamos las librerías necesarias para nuestra red neuronal

import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

In [ ]:
## Creamos nuestro set de datos con algunos números Fibonacci al azar en la lista fibo
## y con su siguiente número Fibonacci en la siguiente lista.

fibo = np.array([8, 55, 144, 610, 4181, 46368, 2178309], dtype=float)
sig_fibo = np.array([13, 89, 233, 987, 6765, 75025, 3524578], dtype=float)

In [ ]:
## Creando nuestra red neuronal, en realidad es solo una neurona 😬

capa = tf.keras.layers.Dense(units=1, input_shape=[1])
model = tf.keras.Sequential([capa])

In [ ]:
model.compile(
    optimizer = tf.keras.optimizers.Adam(0.1),
    loss = 'mean_squared_error'
)

In [ ]:
## Entrenando nuestra neurona 💪🏼

train_model = model.fit(fibo, sig_fibo, epochs=250, verbose=False)
print("Listo!")

In [ ]:
## Veamos como nos va en el entrenamiento

plt.xlabel("Epoch")
plt.ylabel("Pérdida")
plt.plot(train_model.history["loss"])
plt.plot()

In [ ]:
# Predicción de prueba
pred = model.predict([377])
print("El siguiente número en la serie fibonacci de 377 es ", pred)
print("Resultado real 610")

In [ ]:
pred = model.predict([17711])
print("El siguiente número en la serie fibonacci de 17711 es ", str(pred))
print("Resultado real 28657")

In [ ]:
pred = model.predict([55])
print("El siguiente número en la serie fibonacci de 55 es ", str(pred))
print("Resultado real 89")

In [ ]:
print(capa.get_weights())

[array([[1.6180334]], dtype=float32), array([1.2138462], dtype=float32)]


Explicación del modelo

Al ser un modelo de función lineal sigue la forma de la ecuación lineal

$$
y = mx + b
$$

Siendo el peso en este caso de **1.6180334** y su sesgo de **1.2672366**

In [ ]:
# Ahora vamos a mejorar un poco el modelo agregando un par de capas más para ver como se comporta

hidden_1 = tf.keras.layers.Dense(units=3, input_shape=[1])
#hidden_2 = tf.keras.layers.Dense(units=3)
resultado = tf.keras.layers.Dense(units=1)
model = tf.keras.Sequential([hidden_1, resultado])

In [ ]:
model.compile(
    optimizer = tf.keras.optimizers.Adam(0.1),
    loss = 'mean_squared_error'
)

In [ ]:
train_model = model.fit(fibo, sig_fibo, epochs=250, verbose=False)
print("Listo!")

In [ ]:
plt.xlabel("Epoch")
plt.ylabel("Pérdida")
plt.plot(train_model.history["loss"])
plt.plot()

In [ ]:
# Predicción de prueba
pred = model.predict([377])
print("El siguiente número en la serie fibonacci de 377 es ", pred)
print("Resultado real 610")

In [ ]:
pred = model.predict([17711])
print("El siguiente número en la serie fibonacci de 17711 es ", pred)
print("Resultado real 28657")

In [ ]:
pred = model.predict([55])
print("El siguiente número en la serie fibonacci de 55 es ", pred)
print("Resultado real 89")

## Ahora te toca a ti ✌🏻

Bueno ya que hemos visto como funciona esta red neuronal pequeña, ahora te toca hacerla a ti.

Súper simple, te dejaré el mismo código de los número Fibonacci pero debes cambiar los pares por rangos de temperatura fahrenheit y celsius, para luego realizar la predicción de la temperatura.

In [ ]:
## Recuerda entre cada valor separar con una coma

fahren = np.array([], dtype=float)
celsius = np.array([], dtype=float)

In [ ]:
capa = tf.keras.layers.Dense(units=1, input_shape=[1])
model = tf.keras.Sequential([capa])

In [ ]:
model.compile(
    optimizer = tf.keras.optimizers.Adam(0.1),
    loss = 'mean_squared_error'
)

In [ ]:
train_model = model.fit(fahren, celsius, epochs=250, verbose=False)
print("Listo!")

In [ ]:
## Para la predicción pon un valor de temperatura fahrenheit dentro del parentesis cuadrado []

pred = model.predict([]) # --> dentro de los corchetes pon un valor de temperatura fahrenheit
print("En grados celsius es ", pred)